# Capstone Project

## EDA for Car Price Prediction Model

---

In [242]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib as plt
import warnings
import warnings
warnings.filterwarnings('ignore')

In [12]:
df = pd.read_json('scout_car.json', lines=True)

In [9]:
df.sample(20).T

9524   \
url                            https://www.autoscout24.com//offers/opel-corsa...   
make_model                                                            Opel Corsa   
short_description                      1.2 ENJOY +Airco+Cruise Control+Bluetooth   
body_type                                                                 Sedans   
price                                                                       9745   
vat                                                               VAT deductible   
km                                                                     27,941 km   
registration                                                             01/2018   
prev_owner                                                      1 previous owner   
kW                                                                           NaN   
hp                                                                         51 kW   
Type                                                        [, Used, , Gasoline]   
Previous Owners                                                            \n1\n   
Next Inspection                                                              NaN   
Inspection new                                                               NaN   
Warranty                                             [\n12 months\n, \nEuro 6\n]   
Full Service                                                                 NaN   
Non-smoking Vehicle                                                     [\n, \n]   
null                                                                          []   
Make                                                                    \nOpel\n   
Model                                                            [\n, Corsa, \n]   
Offer Number                                                        [\n425163\n]   
First Registration                                                [\n, 2018, \n]   
Body Color                                                       [\n, Black, \n]   
Paint Type                                                        [\nMetallic\n]   
Body Color Original                                                          NaN   
Upholstery                                                    [\nCloth, Black\n]   
Body                                                            [\n, Sedans, \n]   
Nr. of Doors                                                             [\n5\n]   
Nr. of Seats                                                             [\n5\n]   
Model Code                                                                   NaN   
Gearing Type                                                    [\n, Manual, \n]   
Displacement                                                      [\n1,229 cc\n]   
Cylinders                                                                [\n4\n]   
Weight                                                                       NaN   
Drive chain                                                                  NaN   
Fuel                                                          [\n, Gasoline, \n]   
Consumption                                                                  NaN   
CO2 Emission                                                        [[], [], []]   
Emission Class                                                      [\nEuro 6\n]   
\nComfort & Convenience\n      [Air conditioning, Cruise control, Electrical ...   
\nEntertainment & Media\n         [Bluetooth, CD player, MP3, On-board computer]   
\nExtras\n                                                        [Alloy wheels]   
\nSafety & Security\n          [ABS, Central door lock, Driver-side airbag, E...   
description                    [\nAlle DECAIGNY TWEEDEHANDS zijn gecertificee...   
Emission Label                                                               NaN   
Gears                                                                    [\n5\n]   
Country version                            

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15919 entries, 0 to 15918
Data columns (total 54 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   url                            15919 non-null  object 
 1   make_model                     15919 non-null  object 
 2   short_description              15873 non-null  object 
 3   body_type                      15859 non-null  object 
 4   price                          15919 non-null  int64  
 5   vat                            11406 non-null  object 
 6   km                             15919 non-null  object 
 7   registration                   15919 non-null  object 
 8   prev_owner                     9091 non-null   object 
 9   kW                             0 non-null      float64
 10  hp                             15919 non-null  object 
 11  Type                           15917 non-null  object 
 12  Previous Owners                9279 non-null  

In [13]:
df.short_description.describe()

count                                 15873
unique                                10001
top       SPB 1.6 TDI 116 CV S tronic Sport
freq                                     64
Name: short_description, dtype: object

In [14]:
df.short_description.unique()

array(['Sportback 1.4 TDI S-tronic Xenon Navi Klima', '1.8 TFSI sport',
       'Sportback 1.6 TDI S tronic Einparkhilfe plus+music', ...,
       'ELYSEE ENERGY dCi 160 EDC',
       'INITIALE Paris TCe 225 EDC GPF ACC EU6',
       'TCe 225 EDC GPF LIM Deluxe Pano,RFK'], dtype=object)

In [15]:
df.body_type.describe()

count      15859
unique         9
top       Sedans
freq        7903
Name: body_type, dtype: object

In [16]:
df.body_type.unique()

array(['Sedans', 'Station wagon', 'Compact', 'Other', 'Coupe', 'Van',
       'Off-Road', 'Convertible', None, 'Transporter'], dtype=object)

In [18]:
df.body_type.isna().sum()

60

In [23]:
df.price.describe()

count    15919.000000
mean     18019.896727
std       7386.169409
min         13.000000
25%      12850.000000
50%      16900.000000
75%      21900.000000
max      74600.000000
Name: price, dtype: float64

In [27]:
df[df.price<5000].T

6066  \
url                            https://www.autoscout24.com//offers/opel-astra...   
make_model                                                            Opel Astra   
short_description                 1.6 D (CDTI) Start/Stop Sports Tourer Business   
body_type                                                          Station wagon   
price                                                                        255   
vat                                                                         None   
km                                                                      5,563 km   
registration                                                             06/2018   
prev_owner                                                                  None   
kW                                                                           NaN   
hp                                                                        100 kW   
Type                                     [, Used, , Diesel (Particulate Filter)]   
Previous Owners                                                              NaN   
Next Inspection                                                              NaN   
Inspection new                                                               NaN   
Warranty                                                                     NaN   
Full Service                                                                 NaN   
Non-smoking Vehicle                                                          NaN   
null                                                                          []   
Make                                                                    \nOpel\n   
Model                                                            [\n, Astra, \n]   
Offer Number                                                                 NaN   
First Registration                                                [\n, 2018, \n]   
Body Color                                                        [\n, Gold, \n]   
Paint Type                                                                   NaN   
Body Color Original                                                          NaN   
Upholstery                                                                   NaN   
Body                                                     [\n, Station wagon, \n]   
Nr. of Doors                                                             [\n5\n]   
Nr. of Seats                                                             [\n5\n]   
Model Code                                                                   NaN   
Gearing Type                                                    [\n, Manual, \n]   
Displacement                                                      [\n1,598 cc\n]   
Cylinders                                                                [\n4\n]   
Weight                                                            [\n1,403 kg\n]   
Drive chain                                                                  NaN   
Fuel                                       [\n, Diesel (Particulate Filter), \n]   
Consumption                    [[3.8 l/100 km (comb)], [4.5 l/100 km (city)],...   
CO2 Emission                                           [\n101 g CO2/km (comb)\n]   
Emission Class                                                      [\nEuro 6\n]   
\nComfort & Convenience\n      [Air conditioning, Cruise control, Electrical ...   
\nEntertainment & Media\n      [Bluetooth, Hands-free equipment, On-board com...   
\nExtras\n                                                           [Roof rack]   
\nSafety & Security\n          [ABS, Central door lock with remote control, D...   
description                                                                   []   
Emission Label                                                               NaN   
Gears                                                                    [\n6\n]   
Country version                             

In [172]:
df.sample(3).T

14019  \
url                            https://www.autoscout24.com//offers/renault-cl...   
make_model                                                          Renault Clio   
short_description                 dCi90 SS I.Paris Leder R-Link/Navi Bose P.Dach   
body_type                                                                 Sedans   
price                                                                      18087   
vat                                                               VAT deductible   
km                                                                          - km   
registration                                                                 -/-   
prev_owner                                                                  None   
kW                                                                           NaN   
hp                                                                         66 kW   
Type                                      [, New, , Diesel (Particulate Filter)]   
Previous Owners                                                              NaN   
Next Inspection                                                              NaN   
Inspection new                                [\nYes\n, \n109 g CO2/km (comb)\n]   
Warranty                                                                     NaN   
Full Service                                                                 NaN   
Non-smoking Vehicle                                        [\n, \n, \nEuro 6c\n]   
null                                                                          []   
Make                                                                 \nRenault\n   
Model                                                             [\n, Clio, \n]   
Offer Number                                                         [\n68716\n]   
First Registration                                                           NaN   
Body Color                                                       [\n, Black, \n]   
Paint Type                                                        [\nMetallic\n]   
Body Color Original                                            [\nblack pearl\n]   
Upholstery                                             [\nFull leather, Other\n]   
Body                                                            [\n, Sedans, \n]   
Nr. of Doors                                                             [\n5\n]   
Nr. of Seats                                                             [\n5\n]   
Model Code                                                                   NaN   
Gearing Type                                                 [\n, Automatic, \n]   
Displacement                                                      [\n1,461 cc\n]   
Cylinders                                                                    NaN   
Weight                                                            [\n1,308 kg\n]   
Drive chain                                                          [\nfront\n]   
Fuel                                       [\n, Diesel (Particulate Filter), \n]   
Consumption                    [[4.1 l/100 km (comb)], [4.3 l/100 km (city)],...   
CO2 Emission                                           [\n109 g CO2/km (comb)\n]   
Emission Class                                                     [\nEuro 6c\n]   
\nComfort & Convenience\n      [Air conditioning, Armrest, Automatic climate ...   
\nEntertainment & Media\n      [Bluetooth, Digital radio, Radio, Sound system...   
\nExtras\n                      [Alloy wheels, Catalytic Converter, Sport seats]   
\nSafety & Security\n          [ABS, Central door lock, Daytime running light...   
description                    [\n,  , Neues Fahrzeug inkl.19% MWST, Fahrzeug...   
Emission Label                                                   [\n4 (Green)\n]   
Gears                                                                    [\n6\n]   
Country version                            

In [30]:
df.vat.describe()

count              11406
unique                 2
top       VAT deductible
freq               10980
Name: vat, dtype: object

In [31]:
df.vat.unique()

array(['VAT deductible', 'Price negotiable', None], dtype=object)

In [32]:
df.registration.describe()

count     15919
unique       48
top         -/-
freq       1597
Name: registration, dtype: object

In [33]:
df.registration.unique()

array(['01/2016', '03/2017', '02/2016', '08/2016', '05/2016', '03/2016',
       '06/2017', '05/2018', '09/2016', '06/2016', '10/2016', '04/2016',
       '06/2018', '11/2017', '07/2016', '12/2016', '09/2018', '04/2018',
       '07/2017', '02/2017', '11/2016', '01/2018', '05/2017', '04/2017',
       '07/2018', '01/2017', '02/2018', '03/2018', '-/-', '08/2017',
       '10/2017', '08/2018', '09/2017', '12/2017', '04/2019', '12/2018',
       '03/2019', '06/2019', '05/2019', '02/2019', '01/2019', '10/2018',
       '11/2018', '09/2019', '07/2019', '08/2019', '11/2019', '12/2019'],
      dtype=object)

In [34]:
df.prev_owner.describe()

count                 9091
unique                   4
top       1 previous owner
freq                  8294
Name: prev_owner, dtype: object

In [35]:
df.prev_owner.unique()

array(['2 previous owners', None, '1 previous owner', '3 previous owners',
       '4 previous owners'], dtype=object)

In [45]:
df.prev_owner.value_counts()

1 previous owner     8294
2 previous owners     778
3 previous owners      17
4 previous owners       2
Name: prev_owner, dtype: int64

In [47]:
df.prev_owner.isnull().sum()

6828

In [37]:
df['Previous Owners'].describe()

count      9279
unique      102
top       \n1\n
freq       8101
Name: Previous Owners, dtype: object

In [44]:
#df['Previous Owners'].value_counts(dropna=False)

In [51]:
df.Type.describe()

count                                       15917
unique                                        169
top       [, Used, , Diesel (Particulate Filter)]
freq                                         3475
Name: Type, dtype: object

In [52]:
df['Next Inspection'].describe()

count            3535
unique           1399
top       \n04/2022\n
freq               62
Name: Next Inspection, dtype: object

In [53]:
df['Inspection new'].describe()

count                      3932
unique                      202
top       [\nYes\n, \nEuro 6\n]
freq                        523
Name: Inspection new, dtype: object

In [54]:
df.Warranty.describe()

count                    10499
unique                     515
top       [\n, \n, \nEuro 6\n]
freq                      1868
Name: Warranty, dtype: object

In [68]:
df.Warranty.str.contains("CO", case = False).value_counts()

False    3036
Name: Warranty, dtype: int64

In [62]:
df.Warranty.isna().sum()

5420

In [71]:
df.Warranty.dtype.type

numpy.object_

In [72]:
df['Full Service'].describe()

count                        8215
unique                        122
top       [\n, \n, \n4 (Green)\n]
freq                         2235
Name: Full Service, dtype: object

In [78]:
df['Non-smoking Vehicle'].describe()

count         7177
unique          94
top       [\n, \n]
freq          3647
Name: Non-smoking Vehicle, dtype: object

In [82]:
df['Non-smoking Vehicle'].isnull().sum()

8742

In [88]:
df['Make'].describe()

count        15919
unique           3
top       \nOpel\n
freq          7343
Name: Make, dtype: object

In [89]:
df['Make'].unique()

array(['\nAudi\n', '\nOpel\n', '\nRenault\n'], dtype=object)

In [90]:
df['Model'].describe()

count            15919
unique               9
top       [\n, A3, \n]
freq              3097
Name: Model, dtype: object

In [96]:
df['First Registration'].describe()

count              14322
unique                 4
top       [\n, 2018, \n]
freq                4522
Name: First Registration, dtype: object

In [120]:
df['First Registration'].value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[\n, 2018, \n]    4522
[\n, 2016, \n]    3674
[\n, 2017, \n]    3273
[\n, 2019, \n]    2853
Name: First Registration, dtype: int64

In [100]:
df.registration.isin(['2018']).value_counts()

False    15919
Name: registration, dtype: int64

In [101]:
df.registration.value_counts()

-/-        1597
03/2018     695
02/2019     585
05/2018     572
03/2019     543
04/2018     541
01/2019     541
02/2018     539
03/2016     536
04/2016     532
06/2018     532
01/2018     511
04/2019     506
02/2016     472
03/2017     471
05/2016     459
06/2016     452
05/2019     440
06/2017     409
05/2017     404
07/2018     396
04/2017     380
01/2016     376
02/2017     368
01/2017     306
08/2018     285
06/2019     224
07/2017     215
11/2017     180
07/2016     176
10/2016     160
10/2017     154
09/2017     149
11/2016     142
09/2016     141
09/2018     141
12/2016     134
12/2017     123
08/2017     114
11/2018     110
12/2018     103
10/2018      97
08/2016      94
07/2019       6
09/2019       5
12/2019       1
11/2019       1
08/2019       1
Name: registration, dtype: int64

In [108]:
df.registration.isnull().sum()

0

In [109]:
df['Body Color'].describe()

count               15322
unique                 14
top       [\n, Black, \n]
freq                 3745
Name: Body Color, dtype: object

In [118]:
df['Body Color'].value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[\n, Black, \n]     3745
[\n, Grey, \n]      3505
[\n, White, \n]     3406
[\n, Silver, \n]    1647
[\n, Blue, \n]      1431
[\n, Red, \n]        957
[\n, Brown, \n]      289
[\n, Green, \n]      154
[\n, Beige, \n]      108
[\n, Yellow, \n]      51
[\n, Violet, \n]      18
[\n, Bronze, \n]       6
[\n, Orange, \n]       3
[\n, Gold, \n]         2
Name: Body Color, dtype: int64

In [121]:
df['Body Color Original'].describe()

count                  12160
unique                  1927
top       [\nOnyx Schwarz\n]
freq                     338
Name: Body Color Original, dtype: object

In [123]:
df['Paint Type'].describe()

count              10147
unique                 3
top       [\nMetallic\n]
freq                9794
Name: Paint Type, dtype: object

In [127]:
df['Paint Type'].isnull().sum()

5772

In [132]:
df['Upholstery'].value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[\nCloth, Black\n]           5821
[\nPart leather, Black\n]    1121
[\nCloth\n]                  1005
[\nCloth, Grey\n]             891
[\nCloth, Other\n]            639
[\nFull leather, Black\n]     575
[\nBlack\n]                   491
[\nGrey\n]                    273
[\nOther, Other\n]            182
[\nPart leather\n]            140
[\nFull leather\n]            139
[\nFull leather, Brown\n]     116
[\nPart leather, Grey\n]      116
[\nOther, Black\n]            110
[\nFull leather, Other\n]      72
[\nFull leather, Grey\n]       67
[\nPart leather, Other\n]      65
[\nOther\n]                    56
[\nPart leather, Brown\n]      50
[\nalcantara, Black\n]         47
[\nVelour, Black\n]            36
[\nFull leather, Beige\n]      36
[\nCloth, Brown\n]             28
[\nVelour\n]                   16
[\nOther, Grey\n]              15
[\nCloth, Beige\n]             13
[\nBrown\n]                    12
[\nCloth, Blue\n]              12
[\nCloth, White\n]              8
[\nVelour, Gre

In [133]:
df['Nr. of Doors'].describe()

count       15707
unique          6
top       [\n5\n]
freq        11575
Name: Nr. of Doors, dtype: object

In [160]:
df['Nr. of Doors'].value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[\n5\n]    11575
[\n4\n]     3079
[\n3\n]      832
[\n2\n]      219
[\n7\n]        1
[\n1\n]        1
Name: Nr. of Doors, dtype: int64

In [161]:
df['Nr. of Seats'].value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[\n5\n]    13336
[\n4\n]     1125
[\n7\n]      362
[\n2\n]      116
[\n6\n]        2
[\n3\n]        1
Name: Nr. of Seats, dtype: int64

In [162]:
df['Gearing Type'].value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[\n, Manual, \n]            8153
[\n, Automatic, \n]         7297
[\n, Semi-automatic, \n]     469
Name: Gearing Type, dtype: int64

In [166]:
df.Gears.describe()

count       11207
unique         10
top       [\n6\n]
freq         5822
Name: Gears, dtype: object

In [170]:
df.Gears.value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[\n6\n]     5822
[\n5\n]     3239
[\n7\n]     1908
[\n8\n]      224
[\n9\n]        6
[\n4\n]        2
[\n3\n]        2
[\n1\n]        2
[\n50\n]       1
[\n2\n]        1
Name: Gears, dtype: int64

In [171]:
df.Displacement.describe()

count              15423
unique                77
top       [\n1,598 cc\n]
freq                4761
Name: Displacement, dtype: object

In [174]:
df.Cylinders.describe()

count       10239
unique          7
top       [\n4\n]
freq         8105
Name: Cylinders, dtype: object

In [175]:
df.Cylinders.value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[\n4\n]    8105
[\n3\n]    2104
[\n5\n]      22
[\n6\n]       3
[\n2\n]       2
[\n8\n]       2
[\n1\n]       1
Name: Cylinders, dtype: int64

In [176]:
df.Weight.describe()

count               8945
unique               434
top       [\n1,163 kg\n]
freq                 574
Name: Weight, dtype: object

In [178]:
df['Drive chain'].describe()

count            9061
unique              3
top       [\nfront\n]
freq             8886
Name: Drive chain, dtype: object

In [179]:
df['Drive chain'].value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[\nfront\n]    8886
[\n4WD\n]       171
[\nrear\n]        4
Name: Drive chain, dtype: int64

In [180]:
df.Fuel.describe()

count                                     15919
unique                                       77
top       [\n, Diesel (Particulate Filter), \n]
freq                                       4315
Name: Fuel, dtype: object

In [183]:
df['Consumption'].describe()

count                                                 14013
unique                                                  881
top       [[3.9 l/100 km (comb)], [4.1 l/100 km (city)],...
freq                                                    304
Name: Consumption, dtype: object

In [192]:
df['Consumption'].iloc[3]

[['3.8 l/100 km (comb)'], ['4.3 l/100 km (city)'], ['3.5 l/100 km (country)']]

In [193]:
df['CO2 Emission'].describe()

count                         14111
unique                          123
top       [\n120 g CO2/km (comb)\n]
freq                            740
Name: CO2 Emission, dtype: object

In [195]:
df['Emission Class'].describe()

count            12898
unique               7
top       [\nEuro 6\n]
freq             10139
Name: Emission Class, dtype: object

In [196]:
df['Emission Class'].value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[\nEuro 6\n]          10139
[\nEuro 6d-TEMP\n]     1845
[[], [], []]            607
[\nEuro 6c\n]           127
[\nEuro 5\n]             78
[\nEuro 6d\n]            62
[\nEuro 4\n]             40
Name: Emission Class, dtype: int64

In [197]:
df['\nComfort & Convenience\n'].describe()

count                                                 14999
unique                                                 6198
top       [Air conditioning, Electrical side mirrors, Hi...
freq                                                    216
Name: \nComfort & Convenience\n, dtype: object

In [200]:
df['\nEntertainment & Media\n'].describe()

count                                                 14545
unique                                                  346
top       [Bluetooth, Hands-free equipment, On-board com...
freq                                                   1282
Name: \nEntertainment & Media\n, dtype: object

In [201]:
df['\nExtras\n'].describe()

count              12957
unique               659
top       [Alloy wheels]
freq                3245
Name: \nExtras\n, dtype: object

In [203]:
df['\nSafety & Security\n'].describe()

count                                                 14937
unique                                                 4443
top       [ABS, Central door lock, Daytime running light...
freq                                                    538
Name: \nSafety & Security\n, dtype: object

In [204]:
df.description.describe()

count     15919
unique    12874
top          []
freq        636
Name: description, dtype: object

0

### Things Needs to Be Done

1. Think about removing url column.
2. **Get rid of the column "make_model"**
3. **short description
4. Body_type : There are 60 nulls values that we have to deal with.
5. Be aware of outliers such as min. in price column.
6. Vat column : Change it to deductible or not deductible
7. For the km column remove the "km" from the rows and change the data type to int or float.
8. **Drop the registration column because there is another registratio colum which only contains years.**
9. **From prev_owner column remove the part of "previous owner" and change data type to integer. There are 6828 null values.Remove the "Previous owners" column.**
10. **Drop the column "kW"**
11. Change the "hp" column name to kW.Change the data type.
12. Clean the "Type" column.
13. **Drop the columns about inspectation.**
14. Come back to the Warranty column
15. **Drop the Full service column.**
16. **Drop the Non_smoking column.**
17. **Drop the null column.**
18. Clean both make and model columns.
19. Come back to the offer number column again.
20. Change the First registration column name as Registration year
21. **Drop the Body color original**
22. Organize Upholstery column
23. **Drop the Body column.**
24. Check the number of doors
25. For the number od seats, clean '\n'
26. **Drop the model code number  column**
27. Clean the outliers for gearing type column
28. For the Cyclinder column clean the outliers.
29. **You might drop Weight column(!)**
30. Make the small groups for Fuel column.
31. We might need the 'Consumption' to be in 3 different column.
32. For the CO2 Emission, get rid of the unnecessary string parts and change the column name.
33. For the Emission Class column be aware of empty brackets which are null valus and get rid of unnecessary inf.
34. Come back to the \nComfort & Convenience\n column.
35. Come back to '\nEntertainment & Media\n' column.
36. Extras column*
37. Safety & Security column*
38. **Drop the description column.**
39. **Drop the emission column.**
40. Gears column*
41. **Drop rest of the columns.**

In [202]:
df.sample(3).T

6788   \
url                            https://www.autoscout24.com//offers/opel-astra...   
make_model                                                            Opel Astra   
short_description                                      1.6 CDTi Dynamic (110 CV)   
body_type                                                                Compact   
price                                                                      12955   
vat                                                                         None   
km                                                                     37,527 km   
registration                                                             04/2017   
prev_owner                                                                  None   
kW                                                                           NaN   
hp                                                                         81 kW   
Type                                     [, Used, , Diesel (Particulate Filter)]   
Previous Owners                                                              NaN   
Next Inspection                                                              NaN   
Inspection new                                                               NaN   
Warranty                                                           \n12 months\n   
Full Service                                    [\n, \n, \n97 g CO2/km (comb)\n]   
Non-smoking Vehicle                                         [\n, \n, \nEuro 6\n]   
null                                                                          []   
Make                                                                    \nOpel\n   
Model                                                            [\n, Astra, \n]   
Offer Number                                                       [\n8816JYB\n]   
First Registration                                                [\n, 2017, \n]   
Body Color                                                       [\n, Black, \n]   
Paint Type                                                        [\nMetallic\n]   
Body Color Original                                                          NaN   
Upholstery                                                           [\nBlack\n]   
Body                                                           [\n, Compact, \n]   
Nr. of Doors                                                             [\n5\n]   
Nr. of Seats                                                             [\n5\n]   
Model Code                                                                   NaN   
Gearing Type                                                    [\n, Manual, \n]   
Displacement                                                      [\n1,598 cc\n]   
Cylinders                                                                [\n4\n]   
Weight                                                            [\n1,360 kg\n]   
Drive chain                                                          [\nfront\n]   
Fuel                                       [\n, Diesel (Particulate Filter), \n]   
Consumption                    [[3.7 l/100 km (comb)], [4.2 l/100 km (city)],...   
CO2 Emission                                            [\n97 g CO2/km (comb)\n]   
Emission Class                                                      [\nEuro 6\n]   
\nComfort & Convenience\n      [Navigation system, Park Distance Control, Par...   
\nEntertainment & Media\n                                                    NaN   
\nExtras\n                                                        [Alloy wheels]   
\nSafety & Security\n                                                        NaN   
description                    [\n, 8816JYB, Oferta de financiación, consulta...   
Emission Label                                                               NaN   
Gears                                                                    [\n6\n]   
Country version                            

In [177]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15919 entries, 0 to 15918
Data columns (total 54 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   url                            15919 non-null  object 
 1   make_model                     15919 non-null  object 
 2   short_description              15873 non-null  object 
 3   body_type                      15859 non-null  object 
 4   price                          15919 non-null  int64  
 5   vat                            11406 non-null  object 
 6   km                             15919 non-null  object 
 7   registration                   15919 non-null  object 
 8   prev_owner                     9091 non-null   object 
 9   kW                             0 non-null      float64
 10  hp                             15919 non-null  object 
 11  Type                           15917 non-null  object 
 12  Previous Owners                9279 non-null  

In [205]:
df1 = df.copy()

In [206]:
df1.drop('url', axis = 1, inplace = True)
df1.drop('make_model', axis = 1, inplace = True)
df1.drop('short_description', axis = 1, inplace = True)
df1.drop('kW', axis = 1, inplace = True)
df1.drop('registration', axis = 1, inplace = True)
df1.drop('Previous Owners', axis = 1, inplace = True)
df1.drop('Next Inspection', axis = 1, inplace = True)
df1.drop('Inspection new', axis = 1, inplace = True)
df1.drop('Full Service', axis = 1, inplace = True)
df1.drop('null', axis = 1, inplace = True)
df1.drop('Body Color Original', axis = 1, inplace = True)
df1.drop('Body', axis = 1, inplace = True)
df1.drop('Model Code', axis = 1, inplace = True)
df1.drop('Weight', axis = 1, inplace = True)
df1.drop('description', axis = 1, inplace = True)
df1.drop('Emission Label', axis = 1, inplace = True)
df1.drop('Country version', axis = 1, inplace = True)
df1.drop('Electricity consumption', axis = 1, inplace = True)
df1.drop('Last Service Date', axis = 1, inplace = True)
df1.drop('Other Fuel Types', axis = 1, inplace = True)
df1.drop('Availability', axis = 1, inplace = True)
df1.drop('Last Timing Belt Service Date', axis = 1, inplace = True)
df1.drop('Available from', axis = 1, inplace = True)
df1.drop('Non-smoking Vehicle', axis = 1, inplace = True)

In [209]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15919 entries, 0 to 15918
Data columns (total 30 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   body_type                15859 non-null  object
 1   price                    15919 non-null  int64 
 2   vat                      11406 non-null  object
 3   km                       15919 non-null  object
 4   prev_owner               9091 non-null   object
 5   hp                       15919 non-null  object
 6   Type                     15917 non-null  object
 7   Warranty                 10499 non-null  object
 8   Make                     15919 non-null  object
 9   Model                    15919 non-null  object
 10  Offer Number             12744 non-null  object
 11  First Registration       14322 non-null  object
 12  Body Color               15322 non-null  object
 13  Paint Type               10147 non-null  object
 14  Upholstery               12199 non-nul

In [210]:
df1.body_type.describe()

count      15859
unique         9
top       Sedans
freq        7903
Name: body_type, dtype: object

In [211]:
df1.body_type.unique()

array(['Sedans', 'Station wagon', 'Compact', 'Other', 'Coupe', 'Van',
       'Off-Road', 'Convertible', None, 'Transporter'], dtype=object)

In [218]:
df1.km

0        56,013 km
1        80,000 km
2        83,450 km
3        73,000 km
4        16,200 km
           ...    
15914         - km
15915     9,900 km
15916        15 km
15917        10 km
15918         - km
Name: km, Length: 15919, dtype: object

In [262]:
df3 = df.copy()

In [263]:
df3.km

0        56,013 km
1        80,000 km
2        83,450 km
3        73,000 km
4        16,200 km
           ...    
15914         - km
15915     9,900 km
15916        15 km
15917        10 km
15918         - km
Name: km, Length: 15919, dtype: object

In [265]:
df3.km = df3.km.astype('str').replace('- km', np.nan)

In [266]:
df3.km

0        56,013 km
1        80,000 km
2        83,450 km
3        73,000 km
4        16,200 km
           ...    
15914          NaN
15915     9,900 km
15916        15 km
15917        10 km
15918          NaN
Name: km, Length: 15919, dtype: object

In [267]:
df3.km  = df3.km.str.strip(" km")

In [271]:
df3.km = df3.km.str.replace(",", "")

In [272]:
df3.km

0        56013
1        80000
2        83450
3        73000
4        16200
         ...  
15914      NaN
15915     9900
15916       15
15917       10
15918      NaN
Name: km, Length: 15919, dtype: object

In [274]:
df3.km = df3.km.astype(float)

In [275]:
df3.km

0        56013.0
1        80000.0
2        83450.0
3        73000.0
4        16200.0
          ...   
15914        NaN
15915     9900.0
15916       15.0
15917       10.0
15918        NaN
Name: km, Length: 15919, dtype: float64

In [276]:
df1.km = df3.km

In [284]:
df[df.km == "- km"].sample(3).T

14759  \
url                            https://www.autoscout24.com//offers/renault-cl...   
make_model                                                          Renault Clio   
short_description                                     TCe 12V 75 CV 5 porte Life   
body_type                                                                 Sedans   
price                                                                      11200   
vat                                                                         None   
km                                                                          - km   
registration                                                             03/2019   
prev_owner                                                                  None   
kW                                                                           NaN   
hp                                                                         56 kW   
Type                                              [, Pre-registered, , Gasoline]   
Previous Owners                                                              NaN   
Next Inspection                                                              NaN   
Inspection new                                                               NaN   
Warranty                                                           \n24 months\n   
Full Service                                                                 NaN   
Non-smoking Vehicle                                                          NaN   
null                                                                          []   
Make                                                                 \nRenault\n   
Model                                                             [\n, Clio, \n]   
Offer Number                                                      [\n11312976\n]   
First Registration                                                [\n, 2019, \n]   
Body Color                                                       [\n, White, \n]   
Paint Type                                                       [\nUni/basic\n]   
Body Color Original                                                 [\nBianco\n]   
Upholstery                                                    [\nCloth, Black\n]   
Body                                                            [\n, Sedans, \n]   
Nr. of Doors                                                             [\n5\n]   
Nr. of Seats                                                             [\n5\n]   
Model Code                                                                   NaN   
Gearing Type                                                    [\n, Manual, \n]   
Displacement                                                        [\n898 cc\n]   
Cylinders                                                                [\n3\n]   
Weight                                                            [\n1,165 kg\n]   
Drive chain                                                          [\nfront\n]   
Fuel                                                          [\n, Gasoline, \n]   
Consumption                    [[5.2 l/100 km (comb)], [6.5 l/100 km (city)],...   
CO2 Emission                                           [\n118 g CO2/km (comb)\n]   
Emission Class                                                      [\nEuro 6\n]   
\nComfort & Convenience\n      [Air conditioning, Cruise control, Leather ste...   
\nEntertainment & Media\n                           [Bluetooth, MP3, Radio, USB]   
\nExtras\n                                                                   NaN   
\nSafety & Security\n          [ABS, Central door lock, Daytime running light...   
description                    [\nALZACRISTALLI ELETTRICI, CERCHI FLEXWHEEL D...   
Emission Label                                                               NaN   
Gears                                                                    [\n5\n]   
Country version                            

In [301]:
df3.kw

0         66 kW
1        141 kW
2         85 kW
3         66 kW
4         66 kW
          ...  
15914    147 kW
15915    165 kW
15916    146 kW
15917    147 kW
15918    165 kW
Name: kw, Length: 15919, dtype: object

In [305]:
df1.rename(columns={'hp':'k'}, inplace = True )

In [304]:
df1.hp = df3.kW

In [307]:
df3.kw

0         66 kW
1        141 kW
2         85 kW
3         66 kW
4         66 kW
          ...  
15914    147 kW
15915    165 kW
15916    146 kW
15917    147 kW
15918    165 kW
Name: kw, Length: 15919, dtype: object

In [308]:
df1.kW = df3.kw

In [310]:
df1.kW

0         66 kW
1        141 kW
2         85 kW
3         66 kW
4         66 kW
          ...  
15914    147 kW
15915    165 kW
15916    146 kW
15917    147 kW
15918    165 kW
Name: kW, Length: 15919, dtype: object

In [311]:
df1.sample(3).T

5551  \
body_type                                                             Sedans   
price                                                                  22900   
vat                                                           VAT deductible   
km                                                                      50.0   
prev_owner                                                              None   
kW                                                                     85 kW   
Type                                                    [, Used, , Gasoline]   
Warranty                                                       \n24 months\n   
Make                                                                \nAudi\n   
Model                                                           [\n, A3, \n]   
Offer Number                                                    [\n514410\n]   
First Registration                                            [\n, 2019, \n]   
Body Color                                                   [\n, White, \n]   
Paint Type                                                               NaN   
Upholstery                                                               NaN   
Nr. of Doors                                                         [\n5\n]   
Nr. of Seats                                                         [\n5\n]   
Gearing Type                                                [\n, Manual, \n]   
Displacement                                                    [\n999 cc\n]   
Cylinders                                                            [\n3\n]   
Drive chain                                                              NaN   
Fuel                                                      [\n, Gasoline, \n]   
Consumption                [[5 l/100 km (comb)], [5.6 l/100 km (city)], [...   
CO2 Emission                                       [\n118 g CO2/km (comb)\n]   
Emission Class                                                           NaN   
\nComfort & Convenience\n  [Air conditioning, Automatic climate control, ...   
\nEntertainment & Media\n                         [On-board computer, Radio]   
\nExtras\n                     [Alloy wheels, Sport seats, Sport suspension]   
\nSafety & Security\n      [ABS, Driver-side airbag, Electronic stability...   
Gears                                                                [\n6\n]   

                                                                        7328  \
body_type                                                             Sedans   
price                                                                  15270   
vat                                                           VAT deductible   
km                                                                   25350.0   
prev_owner                                                  1 previous owner   
kW                                                                     92 kW   
Type                               [, Used, , Super 95 / Regular/Benzine 91]   
Warranty                                                                 NaN   
Make                                                                \nOpel\n   
Model                                                        [\n, Astra, \n]   
Offer Number                                                     [\nB76-5\n]   
First Registration                                            [\n, 2018, \n]   
Body Color                                                    [\n, Grey, \n]   
Paint Type                                                    [\nMetallic\n]   
Upholstery                                                [\nCloth, Black\n]   
Nr. of Doors                                                         [\n5\n]   
Nr. of Seats                                                         [\n5\n]   
Gearing Type                                                [\n, Manual, \n]   
Displacement                                                  [\n1,399 

In [316]:
df1.kW = df1.kW.str.strip(' kW')

In [317]:
df1.kW

0         66
1        141
2         85
3         66
4         66
        ... 
15914    147
15915    165
15916    146
15917    147
15918    165
Name: kW, Length: 15919, dtype: object

In [325]:
df1.kW.sample(5)

10735    125
14527     56
2266      70
9611      55
6483      70
Name: kW, dtype: object

In [338]:
df1.kW = df1.kW.replace(0,np.nan)

In [339]:
df1.kW = df1.kW.astype(float)

In [340]:
df1.kW

0         66.0
1        141.0
2         85.0
3         66.0
4         66.0
         ...  
15914    147.0
15915    165.0
15916    146.0
15917    147.0
15918    165.0
Name: kW, Length: 15919, dtype: float64

In [341]:
df1.kW.isnull().value_counts()

False    15831
True        88
Name: kW, dtype: int64

In [342]:
df1.prev_owner

0        2 previous owners
1                     None
2         1 previous owner
3         1 previous owner
4         1 previous owner
               ...        
15914                 None
15915     1 previous owner
15916     1 previous owner
15917                 None
15918     1 previous owner
Name: prev_owner, Length: 15919, dtype: object

In [344]:
df1.prev_owner = df1.prev_owner.str.strip(' previous owner')

In [345]:
df1.prev_owner

0           2
1        None
2           1
3           1
4           1
         ... 
15914    None
15915       1
15916       1
15917    None
15918       1
Name: prev_owner, Length: 15919, dtype: object

In [348]:
df1.prev_owner = df1.prev_owner.astype(float)

In [349]:
df1.prev_owner

0        2.0
1        NaN
2        1.0
3        1.0
4        1.0
        ... 
15914    NaN
15915    1.0
15916    1.0
15917    NaN
15918    1.0
Name: prev_owner, Length: 15919, dtype: float64

In [350]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15919 entries, 0 to 15918
Data columns (total 30 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   body_type                15859 non-null  object 
 1   price                    15919 non-null  int64  
 2   vat                      11406 non-null  object 
 3   km                       14895 non-null  float64
 4   prev_owner               9091 non-null   float64
 5   kW                       15831 non-null  float64
 6   Type                     15917 non-null  object 
 7   Warranty                 10499 non-null  object 
 8   Make                     15919 non-null  object 
 9   Model                    15919 non-null  object 
 10  Offer Number             12744 non-null  object 
 11  First Registration       14322 non-null  object 
 12  Body Color               15322 non-null  object 
 13  Paint Type               10147 non-null  object 
 14  Upholstery            

### Things Needs to Be Done

1. Think about removing url column.
2. **Get rid of the column "make_model"**
3. **short description
4. Body_type : There are 60 nulls values that we have to deal with.
5. Be aware of outliers such as min. in price column.
6. Vat column : Change it to deductible or not deductible
7. For the km column remove the "km" from the rows and change the data type to int or float.
8. **Drop the registration column because there is another registratio colum which only contains years.**
9. **From prev_owner column remove the part of "previous owner" and change data type to integer. There are 6828 null values.Remove the "Previous owners" column.**
10. **Drop the column "kW"**
11. Change the "hp" column name to kW.Change the data type.
12. Clean the "Type" column.
13. **Drop the columns about inspectation.**
14. Come back to the Warranty column
15. **Drop the Full service column.**
16. **Drop the Non_smoking column.**
17. **Drop the null column.**
18. Clean both make and model columns.
19. Come back to the offer number column again.
20. Change the First registration column name as Registration year
21. **Drop the Body color original**
22. Organize Upholstery column
23. **Drop the Body column.**
24. Check the number of doors
25. For the number od seats, clean '\n'
26. **Drop the model code number  column**
27. Clean the outliers for gearing type column
28. For the Cyclinder column clean the outliers.
29. **You might drop Weight column(!)**
30. Make the small groups for Fuel column.
31. We might need the 'Consumption' to be in 3 different column.
32. For the CO2 Emission, get rid of the unnecessary string parts and change the column name.
33. For the Emission Class column be aware of empty brackets which are null valus and get rid of unnecessary inf.
34. Come back to the \nComfort & Convenience\n column.
35. Come back to '\nEntertainment & Media\n' column.
36. Extras column*
37. Safety & Security column*
38. **Drop the description column.**
39. **Drop the emission column.**
40. Gears column*
41. **Drop rest of the columns.**

In [352]:
df1.Type

0                  [, Used, , Diesel (Particulate Filter)]
1                                     [, Used, , Gasoline]
2                  [, Used, , Diesel (Particulate Filter)]
3                  [, Used, , Diesel (Particulate Filter)]
4                  [, Used, , Diesel (Particulate Filter)]
                               ...                        
15914               [, New, , Diesel (Particulate Filter)]
15915    [, Used, , Super 95 / Super Plus 98 (Particula...
15916                         [, Pre-registered, , Diesel]
15917                         [, Pre-registered, , Diesel]
15918                        [, Demonstration, , Super 95]
Name: Type, Length: 15919, dtype: object

In [353]:
df3.Model

0            [\n, A1, \n]
1            [\n, A1, \n]
2            [\n, A1, \n]
3            [\n, A1, \n]
4            [\n, A1, \n]
               ...       
15914    [\n, Espace, \n]
15915    [\n, Espace, \n]
15916    [\n, Espace, \n]
15917    [\n, Espace, \n]
15918    [\n, Espace, \n]
Name: Model, Length: 15919, dtype: object

In [356]:
df3.Model[:][1]

['\n', 'A1', '\n']

In [357]:
df3.Make

0           \nAudi\n
1           \nAudi\n
2           \nAudi\n
3           \nAudi\n
4           \nAudi\n
            ...     
15914    \nRenault\n
15915    \nRenault\n
15916    \nRenault\n
15917    \nRenault\n
15918    \nRenault\n
Name: Make, Length: 15919, dtype: object

In [362]:
df2 = df1.copy()

In [369]:
df2.Make = df2.Make.str.strip('\n')

In [371]:
df1.Make = df2.Make

In [372]:
df1.Make

0           Audi
1           Audi
2           Audi
3           Audi
4           Audi
          ...   
15914    Renault
15915    Renault
15916    Renault
15917    Renault
15918    Renault
Name: Make, Length: 15919, dtype: object

In [373]:
df3.Model

0            [\n, A1, \n]
1            [\n, A1, \n]
2            [\n, A1, \n]
3            [\n, A1, \n]
4            [\n, A1, \n]
               ...       
15914    [\n, Espace, \n]
15915    [\n, Espace, \n]
15916    [\n, Espace, \n]
15917    [\n, Espace, \n]
15918    [\n, Espace, \n]
Name: Model, Length: 15919, dtype: object

In [375]:
#for i in df3.Model:
    #print(i[1])

In [380]:
df2.Model.strip('\n')

AttributeError: 'Series' object has no attribute 'strip'

In [381]:
df3.Model

0            [\n, A1, \n]
1            [\n, A1, \n]
2            [\n, A1, \n]
3            [\n, A1, \n]
4            [\n, A1, \n]
               ...       
15914    [\n, Espace, \n]
15915    [\n, Espace, \n]
15916    [\n, Espace, \n]
15917    [\n, Espace, \n]
15918    [\n, Espace, \n]
Name: Model, Length: 15919, dtype: object

In [383]:
df3.Model=df3.Model.str[1].str.strip('\n')

In [384]:
df3.Model

0            A1
1            A1
2            A1
3            A1
4            A1
          ...  
15914    Espace
15915    Espace
15916    Espace
15917    Espace
15918    Espace
Name: Model, Length: 15919, dtype: object

In [385]:
df1.Model = df3.Model

In [386]:
df1.sample(3).T

13387  \
body_type                                                             Sedans   
price                                                                   9900   
vat                                                         Price negotiable   
km                                                                   51000.0   
prev_owner                                                               1.0   
kW                                                                      66.0   
Type                                 [, Used, , Diesel (Particulate Filter)]   
Warranty                                                                 NaN   
Make                                                                 Renault   
Model                                                                   Clio   
Offer Number                                                             NaN   
First Registration                                            [\n, 2017, \n]   
Body Color                                                   [\n, White, \n]   
Paint Type                                                               NaN   
Upholstery                                         [\nPart leather, Black\n]   
Nr. of Doors                                                         [\n5\n]   
Nr. of Seats                                                         [\n5\n]   
Gearing Type                                             [\n, Automatic, \n]   
Displacement                                                  [\n1,461 cc\n]   
Cylinders                                                            [\n4\n]   
Drive chain                                                              NaN   
Fuel                                   [\n, Diesel (Particulate Filter), \n]   
Consumption                [[3.3 l/100 km (comb)], [3.5 l/100 km (city)],...   
CO2 Emission                                        [\n85 g CO2/km (comb)\n]   
Emission Class                                                  [\nEuro 6\n]   
\nComfort & Convenience\n  [Air conditioning, Electrical side mirrors, Le...   
\nEntertainment & Media\n                     [Bluetooth, On-board computer]   
\nExtras\n                                                    [Alloy wheels]   
\nSafety & Security\n      [ABS, Alarm system, Central door lock, Driver-...   
Gears                                                                [\n6\n]   

                                                                       15263  \
body_type                                                                Van   
price                                                                  18480   
vat                                                                     None   
km                                                                  112300.0   
prev_owner                                                               1.0   
kW                                                                     147.0   
Type                                                    [, Used, , Gasoline]   
Warranty                             [\n6 months\n, \n140 g CO2/km (comb)\n]   
Make                                                                 Renault   
Model                                                                 Espace   
Offer Number                                                    [\nc061c3\n]   
First Registration                                            [\n, 2016, \n]   
Body Color                                                   [\n, White, \n]   
Paint Type                                                               NaN   
Upholstery                                                               NaN   
Nr. of Doors                                                         [\n5\n]   
Nr. of Seats                                                         [\n5\n]   
Gearing Type                                             [\n, Automatic, \n]   
Displacement                                                  [\n1,618

In [387]:
df4 = df.copy()

In [394]:
df4.Model = df4.Model.str[1]

In [396]:
df4.Model

0            A1
1            A1
2            A1
3            A1
4            A1
          ...  
15914    Espace
15915    Espace
15916    Espace
15917    Espace
15918    Espace
Name: Model, Length: 15919, dtype: object